retrieval/discovery/consul.go

```diff
@@ -95,6 +96,17 @@ func NewConsulDiscovery(conf *config.ConsulSDConfig) *ConsulDiscovery {
 		scrapedServices: map[string]struct{}{},
 		services:        map[string]*consulService{},
 	}
+	// If the datacenter isn't set in the clientConf, let's get it from the local Consul agent
+	// (Consul default is to use local node's datacenter if one isn't given for a query)
+	if clientConf.Datacenter == "" {
+		info, err := client.Agent().Self()
+		if err != nil {
+			panic(fmt.Errorf("discovery.NewConsulDiscovery: %s", err))
```

Comment:
This error can happen due to an HTTP communication error with the Consul agent. Panics should only be used to catch coding errors within Prometheus, since they will crash the entire server.

Not sure what the right thing to do here is, since it's during the construction of the SD provider. Behavior-wise, I guess periodically retrying to create this SD provider would be desirable if it can't talk to the consul agent (or fall back to omitting the DC label? Meh.), but the calling code isn't really structured to allow for that easily. @fabxc ideas?